libmscore/mmrest.cpp

```diff
@@ -0,0 +1,232 @@
+    _mmWidth = score()->styleP(Sid::minMMRestWidth) * mag();
+    // setbbox(QRectF(0.0, -_spatium, _mmWidth, 2.0 * _spatium));
+    return;
+}
+
+void MMRest::layout(qreal val)
+{
+    _mmWidth = val;
+    qreal vStrokeHeight = score()->styleP(Sid::mmRestHBarVStrokeHeight);
+    bbox().setRect(0.0, -(vStrokeHeight * .5), _mmWidth, vStrokeHeight);
+    addbbox(mmRestNumberRect());
+}
+
+//---------------------------------------------------------
+//   mmRestNumberRect
+///   returns the mmrest number's bounding rectangle
+//---------------------------------------------------------
+
+QRectF MMRest::mmRestNumberRect() const
+{
+    int n = measure()->mmRestCount();
+    std::vector<SymId>&& s = toTimeSigString(QString("%1").arg(n));
+
+    QRectF r = symBbox(s);
+    qreal y = _mmRestNumberPos * spatium() - staff()->height() * .5;
+    qreal x = (_mmWidth - r.width()) * .5;
+
+    r.translate(QPointF(x, y));
+    return r;
+}
+
+//---------------------------------------------------------
+//   MMRest::write
+//---------------------------------------------------------
+
+void MMRest::write(XmlWriter& xml) const
+{
+    xml.stag("Rest");
+    ChordRest::writeProperties(xml);
+    el().write(xml);
+    xml.etag();
+}
+
+//---------------------------------------------------------
+//   propertyDefault
+//---------------------------------------------------------
+
+QVariant MMRest::propertyDefault(Pid propertyId) const
+{
+    switch (propertyId) {
+        case Pid::MMREST_NUMBER_POS:
+            return score()->styleV(Sid::mmRestNumberPos);
+        default:
+            return Rest::propertyDefault(propertyId);
+    }
+}
+
+//---------------------------------------------------------
+//   resetProperty
+//---------------------------------------------------------
+
+void MMRest::resetProperty(Pid propertyId)
+{
+    setProperty(propertyId, propertyDefault(propertyId));
+    return;
+}
+
+//---------------------------------------------------------
+//   getPropertyStyle
+//---------------------------------------------------------
+
+Sid MMRest::getPropertyStyle(Pid propertyId) const
+{
+    if (propertyId == Pid::MMREST_NUMBER_POS) {
+        return Sid::mmRestNumberPos;
+    }
+    return Rest::getPropertyStyle(propertyId);
+}
+
+//---------------------------------------------------------
+//   getProperty
+//---------------------------------------------------------
+
+QVariant MMRest::getProperty(Pid propertyId) const
+{
+    switch (propertyId) {
+        case Pid::MMREST_NUMBER_POS:
+            return _mmRestNumberPos;
+        default:
+            return Rest::getProperty(propertyId);
+    }
+}
+
+//---------------------------------------------------------
+//   setProperty
+//---------------------------------------------------------
+
+bool MMRest::setProperty(Pid propertyId, const QVariant& v)
+{
+    switch (propertyId) {
+        case Pid::MMREST_NUMBER_POS:
+            _mmRestNumberPos = v.toDouble();
+            triggerLayout();
```

Comment:
What about `setLayoutInvalid` Doesn't `trigger` layout compute lots of things?

Reply:
`triggerLayout()` is definitely the right method to use. `setLayoutInvalid()` only applies to elements that are subtypes of `TextBase`. And `triggerLayout()` does not exactly compute lots of things, but it does invoke a number of functions. It is all very fast, though.

Reply:
More specifically, triggerLayout tells MuseScore to relay our exactly the part of the score that this element affects.  So in most cases it means only a single system or page gets laid out after each change rather than the whole score.

setLayoutInvalid basically means, we need to relook at the content of the text element to reprocess all the formatting within in it.